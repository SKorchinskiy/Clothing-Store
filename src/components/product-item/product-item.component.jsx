import {
  ProductItemDetails,
  ProductItemContainer,
} from "./product-item.styles";

import Button from "../button/button.component";

import { cartSelector } from "../../redux/selectors/cart.selector";
import { addToCartAction } from "../../redux/actions/cart/cart.action";
import { useDispatch, useSelector } from "react-redux";

function ProductItem({ product }) {
  const { name, imageUrl, price } = product;

  const dispatch = useDispatch();
  const { cart } = useSelector(cartSelector);

  const addToCart = () => dispatch(addToCartAction(cart, product));

  return (
    <ProductItemContainer>
      <img src={imageUrl} alt={name} />
      <ProductItemDetails>
        <span>{name}</span>
        <span>${price}</span>
      </ProductItemDetails>
      <Button
        type="button"
        value={"Add To Cart"}
        additionalClasses="invert-btn"
        onClick={addToCart}
      />
    </ProductItemContainer>
  );
}

export default ProductItem;
