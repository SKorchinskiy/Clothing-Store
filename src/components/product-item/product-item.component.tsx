import {
  ProductItemDetails,
  ProductItemContainer,
} from "./product-item.styles";

import Button, { ButtonType } from "../button/button.component";

import { selectCartItems } from "../../redux/selectors/cart.selector";
import { addToCartAction } from "../../redux/actions/cart/cart.action";
import { useDispatch, useSelector } from "react-redux";

export type ProductType = {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
};

type ProductItemInputType = {
  product: ProductType;
};

function ProductItem({ product }: ProductItemInputType) {
  const { name, imageUrl, price } = product;

  const dispatch = useDispatch();
  const cartItems = useSelector(selectCartItems);

  const addToCart = () => dispatch(addToCartAction(cartItems, product));

  return (
    <ProductItemContainer>
      <img src={imageUrl} alt={name} />
      <ProductItemDetails>
        <span>{name}</span>
        <span>${price}</span>
      </ProductItemDetails>
      <Button type="button" buttonType={ButtonType.invert} onClick={addToCart}>
        Add To Cart
      </Button>
    </ProductItemContainer>
  );
}

export default ProductItem;
