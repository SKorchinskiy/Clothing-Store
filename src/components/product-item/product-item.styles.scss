.product-item-container {
  display: flex;
  flex-direction: column;
  position: relative;
  align-items: center;
  font-family: "Amatic SC";
  font-size: 24px;

  img {
    width: 100%;
    height: 95%;
    object-fit: cover;
  }

  button {
    width: 90%;
    position: absolute;
    bottom: 10%;
    opacity: 0;
  }

  .product-item-details {
    display: flex;
    width: 100%;
    justify-content: space-between;
  }

  &:hover {
    button {
      opacity: 0.85;
    }

    img {
      opacity: 0.8;
    }
    scale: 1.01;
    transition-duration: 0.3s;
  }
}
