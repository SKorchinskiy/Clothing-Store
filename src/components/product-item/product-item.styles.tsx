import styled from "styled-components";

export const ProductItemDetails = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;

export const ProductItemContainer = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  align-items: center;
  font-family: "Amatic SC";
  font-size: 24px;
  margin: auto;
  width: 200px;

  img {
    width: 100%;
    height: 95%;
    object-fit: cover;
  }

  button {
    width: 90%;
    position: absolute;
    bottom: 10%;
    opacity: 0;
  }

  &:hover {
    button {
      opacity: 0.85;
      transition: 0.3s;

      &:active {
        scale: 0.9;
      }
    }
    img {
      opacity: 0.8;
    }

    scale: 1.01;
    transition-duration: 0.3s;
  }
`;
