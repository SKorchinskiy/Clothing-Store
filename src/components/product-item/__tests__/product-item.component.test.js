import "@testing-library/jest-dom";
import { act, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import renderWithProviders from "../../../utils/test.utils";

import ProductItem from "../product-item.component";

import { productItemStub } from "../../../stubs/product-item.stub";

describe("<ProductItem /> component", () => {
  it("should render UI elements properly", () => {
    const { asFragment } = renderWithProviders(
      <ProductItem product={productItemStub} />
    );
    expect.assertions(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should add product to cart", async () => {
    const { store } = renderWithProviders(
      <ProductItem product={productItemStub} />
    );
    const getCartItems = () => store.getState().cart.cartItems;

    const addToCartButton = screen.getByRole("button");

    expect.assertions(4);

    expect(getCartItems()).toHaveLength(0);

    await act(async () => await userEvent.click(addToCartButton));
    expect(getCartItems()).toHaveLength(1);

    await act(async () => await userEvent.click(addToCartButton));
    expect(getCartItems()).toHaveLength(1);
    expect(getCartItems()[0].quantity).toBe(2);
  });
});
