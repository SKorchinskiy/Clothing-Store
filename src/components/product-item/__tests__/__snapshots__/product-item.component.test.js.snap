// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> component should render UI elements properly 1`] = `
<DocumentFragment>
  <div
    class="sc-grXZZQ jjwuhe"
  >
    <img
      alt="test product name"
      src="productItemUrl"
    />
    <div
      class="sc-bcPKhP cVEhmC"
    >
      <span>
        test product name
      </span>
      <span>
        $10
      </span>
    </div>
    <button
      class="sc-dkjKgF sc-eDnVMP bSMzDe iPyBQU"
      type="button"
    >
      Add To Cart
    </button>
  </div>
</DocumentFragment>
`;
